File: bindings/csharp/Bison/Rmi/Future.cs
// MIT License © 2025 Binary Dice Games
/**
 * @file Future.cs
 * @brief RAII wrapper around `rmi_future_handle`.
 */

namespace Bdg.Bison.Rmi;

/// <summary>
/// RAII wrapper around an <c>rmi_future_handle</c>. Consumed exactly once via
/// <see cref="GetDynamic"/> or <see cref="GetProxy"/>, or discarded with
/// <see cref="Release"/> / by disposing.
/// </summary>
public sealed class Future : IDisposable
{
    private nint _handle;

    internal Future(nint handle) => _handle = handle;

    public void Wait(long timeoutMs = -1)
    {
        RmiException.Check(Native.rmi_future_wait(_handle, timeoutMs), "future.wait");
    }

    public Dynamic GetDynamic()
    {
        RmiException.Check(Native.rmi_future_get_dynamic(ref _handle, out var outHandle), "future.get_dynamic");
        return new Dynamic(outHandle);
    }

    public Proxy GetProxy()
    {
        RmiException.Check(Native.rmi_future_get_proxy(ref _handle, out var outHandle), "future.get_proxy");
        return new Proxy(outHandle);
    }

    public void Release()
    {
        if (_handle != nint.Zero)
        {
            Native.rmi_future_release(_handle);
            _handle = nint.Zero;
        }
    }

    public void Dispose()
    {
        Release();
        GC.SuppressFinalize(this);
    }

    ~Future() => Release();
}

File: bindings/android/bison-lib/src/main/java/com/bdg/bison/rmi/Future.java
// MIT License © 2025 Binary Dice Games
// Permission is hereby granted, free of charge, to use, copy, modify, and
// distribute this file. See the LICENSE file or
// https://opensource.org/licenses/MIT for details.

package com.bdg.bison.rmi;

import com.bdg.bison.Dynamic;
import com.bdg.bison.NativeLibrary;

/**
 * Wraps an {@code rmi_future_handle} (see {@code include/rmi_c.h}) returned
 * by an {@code *Async} call. Consumed exactly once via {@link #getDynamic}
 * or {@link #getProxy} (matching which async call produced it), or
 * discarded with {@link #close} once {@link #waitFor} confirms completion
 * for a call with no result value (e.g. {@code Proxy#setAsync}/{@code
 * clearAsync}).
 */
public final class Future implements AutoCloseable {
  static {
    NativeLibrary.ensureLoaded();
  }

  /** Default timeout (`-1` = the internal RMI default), matching {@code rmi_c.h}. */
  public static final long DEFAULT_TIMEOUT_MS = -1;

  private long handle;

  Future(long handle) {
    this.handle = handle;
  }

  public void waitFor() {
    waitFor(DEFAULT_TIMEOUT_MS);
  }

  public void waitFor(long timeoutMs) {
    nativeWait(handle, timeoutMs);
  }

  /** Consumes this future, expecting a {@code Dynamic} result (e.g. from {@code call}/{@code get}). */
  public Dynamic getDynamic() {
    long result = nativeGetDynamic(handle);
    handle = 0;
    return Dynamic.wrapOwned(result);
  }

  /** Consumes this future, expecting a {@link Proxy} result (e.g. from {@code instantiateAsync}). */
  public Proxy getProxy() {
    long result = nativeGetProxy(handle);
    handle = 0;
    return result == 0 ? null : new Proxy(result);
  }

  @Override
  public void close() {
    if (handle != 0) {
      nativeRelease(handle);
      handle = 0;
    }
  }

  private static native void nativeWait(long handle, long timeoutMs);
  private static native long nativeGetDynamic(long handle);
  private static native long nativeGetProxy(long handle);
  private static native void nativeRelease(long handle);
}
